pkg/roachprod/vm/ibm/provider.go

```diff
@@ -0,0 +1,900 @@
+	"github.com/IBM/networking-go-sdk/transitgatewayapisv1"
+	"github.com/IBM/platform-services-go-sdk/globalsearchv2"
+	"github.com/IBM/platform-services-go-sdk/globaltaggingv1"
+	"github.com/IBM/platform-services-go-sdk/resourcecontrollerv2"
+	"github.com/IBM/platform-services-go-sdk/resourcemanagerv2"
+	"github.com/IBM/vpc-go-sdk/vpcv1"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/vm"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/vm/flagstub"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	ProviderName = "ibm"
+
+	defaultAPIKeyEnvVarPrefix = "IBM"
+
+	defaultMachineType = "cz2-4x8"
+	defaultCPUArch     = vm.ArchS390x
+	defaultCPUFamily   = "IBM zSystem"
+	defaultRemoteUser  = "ubuntu"
+	defaultImageAMI    = "ibm-ubuntu-22-04-4-minimal-s390x-4"
+	defaultNTPServer   = "time.adn.networklayer.com"
+
+	// Default values for volumes
+	defaultSystemVolumeSize = 100
+	systemVolumeSizeRange   = "[100-250]"
+	defaultVolumeType       = "10iops-tier" // 10 IOPS/GB
+	defaultVolumeSize       = 500
+
+	InstanceNotInitialized = "not-initialized"
+	InstanceInvalidStatus  = "invalid-status"
+	InstanceInvalidTags    = "invalid-tags"
+	InstanceNotRochprod    = "not-roachprod"
+)
+
+var (
+	// Errors used in the package
+	ErrKeyNotFound        = fmt.Errorf("no SSH key found")
+	ErrRegionNotSupported = fmt.Errorf("region not supported by provider instance")
+	ErrInvalidZoneFormat  = fmt.Errorf("invalid zone format")
+	ErrImageNotFound      = fmt.Errorf("image not found")
+	ErrImageObsolete      = fmt.Errorf("image is obsolete, and cannot be used")
+	ErrMissingAPIKey      = fmt.Errorf("missing IBM_APIKEY environment variable")
+
+	// defaultZones is the list of availability zones agreed upon with IBM.
+	// This list is only used as a helper for the CLI args, and not to determine
+	// the zones to use for the cluster, as the supported regions are defined in
+	// the each provider's instance.
+	defaultZones = map[string][]string{
+		"ca-tor": {
+			"ca-tor-1",
+			"ca-tor-2",
+			"ca-tor-3",
+		},
+		"br-sao": {
+			"br-sao-1",
+			"br-sao-2",
+			"br-sao-3",
+		},
+	}
+)
+
+// DefaultZones returns the default zones for the provider.
+// The zones are selected based on the geoDistributed flag. If geoDistributed is
+// true, all zones across all regions are returned. If geoDistributed is false,
+// a random zone is selected from a random region.
+func DefaultZones(geoDistributed bool) []string {
+	if geoDistributed {
+		// Return all zones across all regions
+		zones := []string{}
+		for _, regionZones := range defaultZones {
+			zones = append(zones, regionZones...)
+		}
+		return zones
+	}
+
+	// Build the list of regions
+	regions := []string{}
+	for region := range defaultZones {
```

Comment:
For non-geo distributed clusters, we should probably stick with `ca-tor` AZs; RTT between New York and São Paulo is several times that of Toronto.

Reply:
I see your point, but I'm afraid we'll hit the vCPU quota way too quickly if we don't round robin between the two regions.

Reply:
I think that's ok; we can error out and prompt the user to try `br-sao`. This way the user is aware that the preferred region is out of quota; otherwise, it's easy to miss the fact that your cluster is in a remote region, which could lead to wrong assumptions.

Reply:
Went with a `defaultRegionForNonGeoClusters` constant in `pkg/roachprod/vm/ibm/provider.go`  [here](https://github.com/cockroachdb/cockroach/blob/c072939fcd60867f65f48c484763654171174b1d/pkg/roachprod/vm/ibm/provider.go#L65-L68).

Setting it to an empty string will re-enable the random selection of the region in case we want to go this route.